stream_sidebar: Implement stream navigation behavior.

sujalshah-bit opened the pull request:
This pull request is a completion of @emmawitt PR, addresses the following issues discussed on [CZO ](https://chat.zulip.org/#narrow/stream/9-issues/topic/Sending.20a.20message.20after.20clicking.20the.20main.20stream.20name/near/1598434)and implements the corresponding changes:

1. **Stream Navigation Behavior:** Clicking on the name of a stream in the left sidebar now navigates the user to the top topic in the left sidebar view of that stream, rather than displaying an interleaved view. This change aims to reduce confusion, especially for new users.

2. **Interleaved Button Addition:** An "interleaved" button has been added to the stream header row in the left sidebar. This button appears only when the user hovers over it, similar to the interleaved button on the DMs row. This addition provides users with the option to switch to an interleaved view if desired.
3. **Test Corrections**: Made necessary adjustments in message-basics.test.ts and navigation.test.ts to address failing tests.

## Related Issues:
#18549, #26409

Fixes #26937.

## Top topic on click

| Dark | Light|
|--------|-------|
| ![top topic on click](https://github.com/zulip/zulip/assets/73663475/ba87fb42-cea6-4382-9072-82cccfeb4adb) | ![top topic on click light](https://github.com/zulip/zulip/assets/73663475/95b046fc-b227-4802-9a49-df5b78dc11bb) |

## Show interleave button on hover

| Dark(on hover) | Light(on hover) |
|--------|-------|
| ![interleaved icon on  hover](https://github.com/zulip/zulip/assets/73663475/6f2787bd-36d8-445d-8221-057586e79a36) | ![inter hover light](https://github.com/zulip/zulip/assets/73663475/4f07c145-14fd-44fb-8d1b-b9d502306c9f) |

## Simple Row 

| Dark | Light |
|--------|-------|
| ![simple row header](https://github.com/zulip/zulip/assets/73663475/eae74717-61bc-4934-a6a0-c256b65de968) | ![inter not hover light](https://github.com/zulip/zulip/assets/73663475/5f0b833b-dff1-4a48-b118-30b4d3871c5c) |


## stream with no topics or all muted topics
| no topic  stream view| all muted topic stream view |
|--------|-------|
| ![image](https://github.com/zulip/zulip/assets/73663475/956e88d0-5920-4f18-a294-aa0079fa50ec) | ![image](https://github.com/zulip/zulip/assets/73663475/02e1a044-e852-4794-adb5-9679f51b2c86) |

## Tooltip
| Interleaved button tooltip |
|--------|
| ![image](https://github.com/zulip/zulip/assets/73663475/6ea5b292-4bc1-401e-96db-95be37574afd) |

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sujalshah-bit commented:
@amanagr, I'm requesting your review as you are the owner of the [issue](https://github.com/zulip/zulip/issues/26937).

amanagr commented on web/styles/left_sidebar.css:
```diff
@@ -972,6 +972,22 @@ li.top_left_scheduled_messages {
            definition. */
         height: auto;
     }
+
+    .fa {
+        color: var(--color-left-sidebar-navigation-icon);
+    }
+
+    .fa:hover {
+        color: var(--color-text-message-header);
+    }
+}
+
+.bottom_left_row .fa-align-right {
+    display: none;
+}
+
+.bottom_left_row:hover .fa-align-right {
+    display: inline-block;
```
We only want to apply css for `.stream-markers-and-controls .fa-align-right`. Let's directly use it. Also, you can use `&:hover` to shorten your css.

sujalshah-bit replied:
If we use `.stream-markers-and-controls .fa-align-right`, the interleaved button will only appear when the user hovers over the unread counter, and it will not appear when the stream is hovered over.

**Stream hover:**
![image](https://github.com/zulip/zulip/assets/73663475/1ac922b6-faaf-4a52-9c8f-ae6a6d1cb3c4)

**Unread counter hover:**
![image](https://github.com/zulip/zulip/assets/73663475/91b0e9a4-93f0-4641-b7ac-85c064064f77)

```css

    .fa-align-right {
        color: var(--color-left-sidebar-navigation-icon);
        display: none;
    }
    
    &:hover {
        color: var(--color-text-message-header);
        .fa-align-right {
            display: inline-block;
        }
    }
}

```

amanagr commented on web/e2e-tests/message-basics.test.ts:
```diff
@@ -310,8 +321,8 @@ async function expect_all_direct_messages(page: Page): Promise<void> {
     console.log("Narrowing with left sidebar");
 
-    await page.click((await get_stream_li(page, "Verona")) + " a");
-    await expect_verona_stream(page);
+    await page.click((await get_stream_li(page, "Verona")) + " .stream-name");
+    await expect_verona_stream_top_topic(page);
```
Instead of modifying what this test is checking here, can we go to the interleaved view and check the same thing. Then, add another test which just checks that we are going to the top topic in verona on clicking the stream name from left sidebar.

sujalshah-bit replied:
Just want to clarify if I understood it right.
Are you suggesting that instead of altering `expect_verona_stream`, we navigate to the interleaved view and verify the same condition ( **what same condition are referring to** ) there? then,  adding a separate test to ensure that clicking the stream name from the left sidebar takes us to the top topic in Verona?

amanagr replied:
Yeah, basically what the updated code here does.

amanagr commented on web/src/stream_list.ts:
```diff
@@ -835,6 +838,28 @@ export function set_event_handlers({
+            (topic) => !user_topics.is_topic_muted(stream_id, topic),
+        );
+
+        if (top_unmuted_topic) {
```
We should still narrow to stream instead of doing nothing if there no unmuted topics.

amanagr replied:
This was not correctly addressed to fixed it.

amanagr commented:
@sujalshah-bit looks great! Left some comments. Thanks for working on this.

alya commented:
@amanagr flagged to me offline that we should think about the behavior if all topics in a stream are muted, or there are no topics at all (all messages have been deleted). I think as an initial approach, we can go to the overall stream view in that case.

This would apply just if all topics have been explicitly muted. So in a muted stream with default topic visibility, we should go to the first topic.

alya commented:
Let's add a tooltip to the "interleaved" button, with the same delay as on the "All direct messages" button. We can test out "Stream feed" for the tooltip, in line with #27802.

alya commented:
Sometimes clicking on a stream in the left sidebar doesn't take me anywhere. E.g., it happened when I clicked on # Rome  (not the interleaved button) here:

<img width="335" alt="Screenshot 2024-03-12 at 5 19 44 PM" src="https://github.com/zulip/zulip/assets/2090066/d014b084-54b4-4868-9255-c94034778a38">

# Rome had several topics in it:

<img width="335" alt="Screenshot 2024-03-12 at 5 19 44 PM" src="https://github.com/zulip/zulip/assets/2090066/37298622-024c-4aee-a261-2b6de9b6ed63">

After some clicking around, clicking on # Rome worked again, so I'm not sure what the reproducer is. But I observed this more than once, and once I got into this state, it was broken until I did something else.

alya commented:
I'm very excited about progress on this!

sujalshah-bit commented on web/src/stream_list.ts:
```diff
@@ -822,7 +825,7 @@ export function set_event_handlers({
     on_stream_click: (stream_id: number, trigger: string) => void;
 }): void {
-    $("#stream_filters").on("click", "li .subscription_block", (e) => {
+    $("#stream_filters").on("click", "li .subscription_block a", (e) => {
```
using `.fa-align-right` instead of `a` now.

sujalshah-bit commented:
Updated the PR:

- Added a tooltip to the Interleaved Button. You can view the screenshot in the PR description.

>"This would apply only if all topics have been explicitly muted. So, in a muted stream with default topic visibility, we should go to the first topic."
- Made the necessary changes to address this issue. screenshot added in the PR description.

@amanagr left a comment above regarding test.

alya commented:
> Sometimes clicking on a stream in the left sidebar doesn't take me anywhere. E.g., it happened when I clicked on # Rome (not the interleaved button) here:

What's the state of debugging this issue?

sujalshah-bit commented:
@alya I haven't been able to replicate the bug you mentioned. I've tried playing around with it, but I couldn't find the issue. Could you please check again to see if you're still encountering that bug?

@amanagr I have updated the PR according to your feedbacks Including the tests.

amanagr commented on web/styles/left_sidebar.css:
```diff
@@ -972,6 +972,26 @@ li.top_left_scheduled_messages {
         height: auto;
     }
+
+    .fa-align-right {
+        color: var(--color-left-sidebar-navigation-icon);
+    }
+
+    &:hover {
+        color: var(--color-text-message-header);
+    }
+}
```
Edited to not apply the styles to `.topic-markers-and-controls`.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -974,6 +974,28 @@ li.top_left_scheduled_messages {
+}
+
+.bottom_left_row {
+    .fa-align-right {
```
Let's add a class `goto-interleaved-stream-view` for this element, rather than referring to it by what icon we picked for it everywhere.

sujalshah-bit commented:
@timabbott Updated the PR with new changes kindly review:)

timabbott commented on web/src/stream_list.ts:
```diff
@@ -822,13 +825,51 @@ export function set_event_handlers({
+        );
+        const muted_topics = topics.find(
+            (topic) => !user_topics.is_topic_unmuted(stream_id, topic),
+        );
```
So I think `muted_topics` here is wrong; can you talk about why you calculate it that way?

sujalshah-bit replied:
Sorry for the confusion. I didn't realize that `muted_topics `still contains the **top topic**. Instead, it should be undefined when there are no muted topics."

timabbott commented on web/src/stream_list.ts:
```diff
@@ -822,13 +825,51 @@ export function set_event_handlers({
+        } else {
+            on_stream_click(stream_id, "sidebar");
+            return;
+        }
```
So I think long-term, we're going to need to figure out better error handling for the case that we don't have anything in `stream_topic_history`; but going to the stream seems OK for testing.

Why do you `return` here, though?

sujalshah-bit replied:
I think we don't need else block.

timabbott replied:
No, as noted before, we need to go to the full-stream narrow if a stream contains no topics. I just reproduced it being broken with an empty stream.

Note that empty stream is not the only scenario where this may come up -- if we don't have any messages in the stream fetched in our local cache, this existing logic would also fail.

That second issue be patched by doing `stream_topic_history_util.get_server_history` with an appropriate callback, only for the case that we don't have data in `stream_topic_history.get_recent_topic_names`.

timabbott replied:
I can test-deploy this with the following patch applied, also pushed it here.

``` diff
diff --git a/web/src/stream_list.ts b/web/src/stream_list.ts
index 06262de9dd..a14e07cb8f 100644
--- a/web/src/stream_list.ts
+++ b/web/src/stream_list.ts
@@ -876,6 +876,8 @@ export function set_event_handlers({
         } else if (muted_topics) {
             destination_url = hash_util.by_stream_topic_url(stream_id, muted_topics);
             browser_history.go_to_location(destination_url);
+        } else {
+            on_stream_click(stream_id, "sidebar");
         }
 
         popovers.hide_all();
```

sujalshah-bit commented:
@timabbott this PR is waiting for the review . Posted comment above kindly check :)

sujalshah-bit commented:
@zulipbot  add "buddy:review"

sujalshah-bit commented:
@zulipbot add "buddy review"

adnan-td commented:
Great work! During manual testing noted:

1. In light mode, hovering over the interleave button does not have the hover effect as in dark mode. (referred the  interleave button for DM feed).
![image](https://github.com/zulip/zulip/assets/78212328/12c05007-8d14-4597-ad05-de446b756c4b)
2. Not sure why the tool tip on interleaved button was not showing.

alya commented:
Testing this PR, keyboard navigation with arrows between tabs doesn't seem to be working. (It works with the tabs in the channels menu.)

sujalshah-bit commented:
@alya, can you tell me which tabs? And also, I made the changes on @adnan-td's feedbacks.

sujalshah-bit commented:
![image](https://github.com/zulip/zulip/assets/73663475/d751f376-fdf2-4073-9231-bd5cb0304960)

alya commented:
@adnan-td Is this PR ready for mentor review?

adnan-td commented:
Looks good to me!

adnan-td commented:
@zulipbot remove "buddy review"

adnan-td commented:
@zulipbot add "mentor review"

yogesh-sirsat reviewed (commented):
LGTM!

alya commented:
Works for me in manual testing!

There is some icon colors inconsistency between all our left sidebar buttons, but that's pre-existing, so not a blocker to move forward.

alya commented:
@timabbott This has been buddy and mentor approved. Previously reviewed by @amanagr as well, so I think you can take a look next.

timabbott commented on web/src/stream_list.ts:
```diff
@@ -837,13 +839,46 @@ export function set_event_handlers({
     on_stream_click: (stream_id: number, trigger: string) => void;
 }): void {
-    $("#stream_filters").on("click", "li .subscription_block", (e) => {
-        if (e.metaKey || e.ctrlKey || e.shiftKey) {
```
The previous logic checked `shiftKey` too; why the change?

sujalshah-bit replied:
there was a typo from my side.

sujalshah-bit commented:
@timabbott  made the changes.

timabbott commented:
I don't see any response to the discussion about whether we need to use  `stream_topic_history_util.get_server_history` in https://github.com/zulip/zulip/pull/29191#discussion_r1645212648. Maybe you can comment on your thinking on that here, since it's hidden in the GitHub scrollback that you have to click to see:

>  if we don't have any messages in the stream fetched in our local cache, this existing logic would also fail.
>
> That second issue be patched by doing `stream_topic_history_util.get_server_history` with an appropriate callback, only for the case that we don't have data in `stream_topic_history.get_recent_topic_names`.

sujalshah-bit commented:
Regarding this [comment](https://github.com/zulip/zulip/pull/29191#discussion_r1645212648):

I have made the changes according to your feedback in the comment. Now The click event handler checks if local topics are empty and fetches server history if needed. If topics are still empty, it navigates to the full-stream view; otherwise, it navigates to the appropriate topic. This ensures handling of both empty streams and those without locally cached messages.

Now if there is no topic / no cache, then it will show this:
![image](https://github.com/zulip/zulip/assets/73663475/31bf34bb-aa94-41bc-9ff5-e64fc6315dc6)


I apologize for the for not responding. Your comments were displayed in the upper portion of the screen, which caused me to overlook them unintentionally. I regret that this oversight resulted in an inconvenience for you.

sujalshah-bit commented:
@timabbott can you look :)

sujalshah-bit commented on web/src/stream_list.ts:
```diff
@@ -837,14 +840,67 @@ export function set_event_handlers({
 
-        clear_and_hide_search();
+            popovers.hide_all();
+            clear_and_hide_search();
```
This was unintentional. I've revised the code to ensure that `clear_and_hide_search();` is consistently called in both scenarios—whether topics are fetched from the server or not. Thank you for catching that oversight!

timabbott commented:
> I apologize for the for not responding. Your comments were displayed in the upper portion of the screen, which caused me to overlook them unintentionally. I regret that this oversight resulted in an inconvenience for you.

Yeah I've actually been having the same problem with that thread being lost above; definitely GitHub's fault for the way it organizes the comment threads.

sujalshah-bit commented:
@timabbott is this PR need another review ?

amanagr commented:
Pushed to add user setting to configure this option as per #30565.

<img width="782" alt="Screenshot 2024-07-03 at 4 13 23 PM" src="https://github.com/zulip/zulip/assets/25124304/32e2abc9-533a-42af-a995-98da2d5e59e7">

I am not sure @alya what do you mean this in #30565:

>  Do not show the channel feed button in the left sidebar, regardless of how this option is configured.

alya commented:
We will no longer show the interleaved view button on hover in the left sidebar. If you set Zulip to take you to the first topic, you'll be able to navigate to the channel feed from the three-dot channel menu, but not from the sidebar directly.

amanagr commented:
> We will no longer show the interleaved view button on hover in the left sidebar. If you set Zulip to take you to the first topic, you'll be able to navigate to the channel feed from the three-dot channel menu, but not from the sidebar directly.

Okay, we can add that in a separate PR.

timabbott commented:
I think we should make that change in this PR; since this PR is the one that introduces the interleaved button concept that we want to remove in the first place, I think it makes sense to just place that option in the stream menu as proposed in the spec. It could be valid to merge this with no option to access the interleaved view at all as an intermediate state, but I think we need the option before sending this to Cloud, so I think it'd be best to just implement the design end-to-end in this PR.

Note that for spectators, this would mean we should make the stream menu exist at all; that detail I'm OK with doing in a near-term follow-up, if it somehow is complicated to have a stream menu appear, though I'm not aware of a reason it would be.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+* [`POST /register`](/api/register-queue), [`PATCH /settings`](/api/update-settings),
+  [`PATCH /realm/user_settings_defaults`](/api/update-realm-user-settings-defaults):
+  Added new user setting `web_channel_default_view`. Clients may use this to
+  determine which narrow to show on clicking a channel link in a UI component like
```
"which view", not "which narrow"; we're phasing out the "narrow" term, and also we should keep in mind that some things, like a "channel inbox", would not even be "narrows/message views" technically.

amanagr replied:
Thanks, noted and changed that here.

timabbott commented on zerver/models/users.py:
```diff
@@ -116,6 +116,20 @@ class UserBaseSettings(models.Model):
+        NARROW_TO_CHANNEL_FEED,
+    ]
+
+    web_channel_default_view = models.SmallIntegerField(default=NARROW_TO_FIRST_TOPIC_IN_CHANNEL)
```
Maybe worth checking in #backend, but I think we should use the `db_default` option to make the migration less disruptive. See `git grep db_default` for an existing example.

amanagr replied:
Started discussion in https://chat.zulip.org/#narrow/stream/3-backend/topic/db_default.20vs.20default.20for.20Django.20models

timabbott commented on web/templates/stream_sidebar_row.hbs:
```diff
@@ -8,9 +8,10 @@
+            <span title="{{name}}" class="stream-name">{{name}}</span>
 
             <div class="stream-markers-and-controls">
+                <a href="{{url}}" id="stream-feed" data-tooltip-template-id="stream-feed-template" class="fa fa-align-right goto-interleaved-stream-view"></a>
```
Oh, I see, you didn't understand this part:

> > I am not sure @alya what do you mean this in #30565:

> Do not show the channel feed button in the left sidebar, regardless of how this option is configured.

What was meant was "The new design does not call for this channel feed button" -- so we should delete this line and the associated tooltip code, in favor of an item in the stream's popover menu that's present if your setting is such that the channel feed isn't the default click action.

amanagr replied:
Done.

<img width="503" alt="Screenshot 2024-07-08 at 3 40 46 PM" src="https://github.com/zulip/zulip/assets/25124304/1aced5e1-f6a1-4c08-9e38-0cc245e11d18">

timabbott commented on zerver/models/users.py:
```diff
@@ -116,6 +116,20 @@ class UserBaseSettings(models.Model):
+    # in the channel.
+
+    NARROW_TO_FIRST_TOPIC_IN_CHANNEL = 1
+    NARROW_TO_CHANNEL_FEED = 2
```
Let's call these something more prefixed, like `WEB_CHANNEL_DEFAULT_VIEW_FIRST_TOPIC` and `WEB_CHANNEL_DEFAULT_VIEW_CHANNEL_FEED`.

Later we'll convert these to formal enums to get rid of the really long names, but our current pattern is to namespace these with the setting name.

amanagr replied:
Makes sense, done!

alya commented:
> I think it makes sense to just place the "Channel feed" option in the stream menu as proposed in the spec. It could be valid to merge this with no option to access the interleaved view at all as an intermediate state, but I think we need the option before sending this to Cloud, so I think it'd be best to just implement the design end-to-end in this PR.

I think this means actually finishing #30565 on this PR.

amanagr commented:
Channel popover for spectators.
<img width="503" alt="Screenshot 2024-07-08 at 3 35 12 PM" src="https://github.com/zulip/zulip/assets/25124304/aac698c5-a070-4048-999c-abd9dcbae6f9">

alya commented:
Looks good to me in manual testing!

I'm not sure where the topic should go in the current menu, but I'm happy to leave it be (it's the top link in the current PR) and figure it out in https://github.com/zulip/zulip/issues/30529.

timabbott commented on web/src/ui_init.js:
```diff
@@ -548,7 +548,6 @@ export function initialize_everything(state_data) {
                 ],
                 {trigger},
             );
-            activity_ui.build_user_sidebar();
```
@evykassirer FYI, since this line was introduced in 231aa098cb19cec028a189dded01ea9191021e40. I agree I don't see any function it can have had running here.

evykassirer replied:
Ah, I probably just didn't realize it was also being called in `message_view.show`. There are a few early returns in `show` but I _think_ none of those are cases we'd need to be concerned about, though I didn't look that closely.

timabbott commented on web/templates/stream_sidebar_row.hbs:
```diff
@@ -11,8 +11,6 @@
 
             <div class="stream-markers-and-controls">
-                <a href="{{url}}" id="stream-feed" data-tooltip-template-id="stream-feed-template" class="fa fa-align-right goto-interleaved-stream-view"></a>
-                <span class="unread_mention_info"></span>
```
Did you intend to remove `unread_mention_info`? I think the answer must be no.

timabbott commented:
This is great, thanks for all the work on this @sujalshah-bit and @amanagr. I made a bunch of changes:
- Renumbered the API migration and feature level since I'd just merged another PR.
- Cleaned up the API documentation to be more concise and precise, aligning with some of the wording patterns in the other web setting merged today.
- Fixed https://github.com/zulip/zulip/pull/29191#discussion_r1669244282 to not remove the mention indicator by accident.
- Reordered the commits to have a more readable history that doesn't add code and then delete it 3 commits later. I mostly did this to make reading/reviewing easier, and am glad I did, since doing so is how I caught the mention bug.

Marked to merge once CI passes! Very excited about this change.
